pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleTargetSelector.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.rule;
+
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import net.sourceforge.pmd.Rule;
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.ast.RootNode;
+import net.sourceforge.pmd.lang.rule.internal.TargetSelectorInternal;
+import net.sourceforge.pmd.lang.rule.internal.TreeIndex;
+import net.sourceforge.pmd.util.CollectionUtil;
+
+/**
+ * A strategy for selecting nodes that will be targeted by a rule.
+ *
+ * @see Rule#getTargetSelector()
+ */
+public abstract class RuleTargetSelector extends TargetSelectorInternal {
```

Comment:
Yes this is super weird. I wanted to make RuleTargetSelector published because it's somewhere on the Rule interface, but still keep everything about TreeIndex internal... So no one can implement their RuleTargetSelector because that would require us to expose TreeIndex.

I'm truly not sure this is the best way to do it, if you have a better idea go ahead